File: Programmers/[67256] 키패드 누르기/ONE/67256.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    static class Key {
        int r;
        int c;

        public Key(int r, int c) {
            this.r = r;
            this.c = c;
        }

        public void moveFinger(Key key) {
            this.r = key.r;
            this.c = key.c;
        }

        public int getDistance(Key key) {
            return Math.abs(this.r - key.r) + Math.abs(this.c - key.c);
        }
    }

    static class Hand {
        Key leftFinger;
        Key rightFinger;
        String handed;

        public Hand(String handed) {
            this.leftFinger = new Key(3, 0);
            this.rightFinger = new Key(3, 2);
            this.handed = handed;
        }

        public boolean isRightHanded() {
            return this.handed.equals("right");
        }
    }

    static Map<Integer, Key> keypad = new HashMap<>();

    public String solution(int[] numbers, String hand) {
        StringBuilder answer = new StringBuilder();
        Hand h = new Hand(hand);

        setKeypad();

        for (int num : numbers)
            move(num, h, answer);

        return answer.toString();
    }

    private void setKeypad() {
        keypad.put(0, new Key(3, 1));
        keypad.put(1, new Key(0, 0));
        keypad.put(2, new Key(0, 1));
        keypad.put(3, new Key(0, 2));
        keypad.put(4, new Key(1, 0));
        keypad.put(5, new Key(1, 1));
        keypad.put(6, new Key(1, 2));
        keypad.put(7, new Key(2, 0));
        keypad.put(8, new Key(2, 1));
        keypad.put(9, new Key(2, 2));
    }

    private void move(int num, Hand hand, StringBuilder builder) {
        // Right Side
        if (num % 3 == 0 && num != 0) {
            hand.rightFinger.moveFinger(keypad.get(num));
            builder.append("R");
        }

        // Left Side
        else if (num % 3 == 1) {
            hand.leftFinger.moveFinger(keypad.get(num));
            builder.append("L");
        }

        // Center
        else {
            if (hand.rightFinger.getDistance(keypad.get(num))
                    < hand.leftFinger.getDistance(keypad.get(num))) {
                hand.rightFinger.moveFinger(keypad.get(num));
                builder.append("R");
            }

            else if (hand.rightFinger.getDistance(keypad.get(num))
                    > hand.leftFinger.getDistance(keypad.get(num))) {
                hand.leftFinger.moveFinger(keypad.get(num));
                builder.append("L");
            }

            else {
                if (hand.isRightHanded()) {
                    hand.rightFinger.moveFinger(keypad.get(num));
                    builder.append("R");
                }

                else {
                    hand.leftFinger.moveFinger(keypad.get(num));
                    builder.append("L");
                }
            }
        }
    }
}

File: Programmers/[67256] 키패드 누르기/JFe/67256.py
def solution(numbers, hand):
    answer = ''
    left, right = [0, 3], [2, 3]    # '*', '#' 좌표[x, y] 저장
    for n in numbers:
        if n == 0: n = 11   # 0은 계산 편하게 11로 저장
        x, y = (n-1)%3, (n-1)//3    # 해당 숫자의 키패드 상 좌표 계산
        # 키패드 위치에 따라 나눠서 처리
        if n == 1 or n == 4 or n == 7:
            answer += 'L'
            left = [x, y]
        elif n == 3 or n == 6 or n == 9:
            answer += 'R'
            right = [x, y]
        else:
            # 왼손, 오른손 거리 계산
            left_dir = abs(left[0]-x) + abs(left[1]-y)
            right_dir = abs(right[0]-x) + abs(right[1]-y)
            if left_dir < right_dir:
                answer += 'L'
                left = [x,y]
            elif left_dir > right_dir:
                answer += 'R'
                right = [x, y]
            # 거리가 같으면 hand 값에 따라서 처리
            else:
                if hand == "left": 
                    answer += 'L'
                    left = [x, y]
                else: 
                    answer += 'R'
                    right = [x, y]
    return answer